### PR TITLE
Implement custom arena behavior

### PR DESCRIPTION
**Pull request by fitzgen**
WIP fix for #1

@matklad what do you think of this? Would this enable the use cases you're thinking of? By customizing `ArenaBehavior::new_arena_id`, and adding some missing impls, we could also make arenas compare for equality.

### CODE REVIEW

**Comment by matklad** [comment]
LGTM!

A couple of comments:

There's an inconsistency in how `Arena` and `ArenaBehavior` accept indices: one uses `&`, the other by value. 

In theory, we can make `T` an associated type of `Behavior` as well, so that `Arena` has a single type parameter. However, we'll have to define some kind of type alias then for the default case: `type DefaultArena = Arena<DefaultBehavior<T>>;`

**Comment by matklad** [comment]
Crazy idea: could we extend `ArenaBehavior` to subsume `generational-arena` as well?

**Comment by fitzgen** [comment]
> There's an inconsistency in how `Arena` and `ArenaBehavior` accept indices: one uses `&`, the other by value.

Yeah, I think it makes sense to always take by value, since identifiers are intended to be cheap copy types.

> In theory, we can make `T` an associated type of `Behavior` as well, so that `Arena` has a single type parameter. However, we'll have to define some kind of type alias then for the default case: `type DefaultArena = Arena<DefaultBehavior<T>>;`

I thought about this, but figured that the 99% use case is when you aren't customizing behavior, and in this case, I think it is less cognitive overhead to just say `Arena<T>` rather than `DefaultArena<T>`.  Thoughts?

**Comment by fitzgen** [comment]
Also, do you want to be a co-maintainer of this crate?

**Comment by matklad** [comment]
>Yeah, I think it makes sense to always take by value, since identifiers are intended to be cheap copy types.

+1

>I think it is less cognitive overhead to just say Arena<T> rather than DefaultArena<T>. 

+1

>Also, do you want to be a co-maintainer of this crate?

Sure: hopefully, it won't need a lot of maintenance!

**Comment by fitzgen** [comment]
> Crazy idea: could we extend `ArenaBehavior` to subsume `generational-arena` as well?

Interesting idea... perhaps? I don't see any reason why not, assuming that we add the appropriate hooks to `ArenaBehavior`.

I'll focus on getting this PR into shape to land, and then maybe you could poke at it and see if you could get a proof of concept for extending the trait to handle generational-arena as well?

> Sure: hopefully, it won't need a lot of maintenance!

Indeed :)

**Comment by matklad** [comment]
@fitzgen could you perhaps add `impl ParitalEq/Eq` for arena in this PR, so that I can test it with rust-analizer right away?

**Comment by fitzgen** [comment]
Sure thing

**Comment by fitzgen** [comment]
Implemented!

File: src/lib.rs [hunk 1]
```diff
@@ -119,6 +117,63 @@ mod imports {
...
+
+    /// Construct a new object identifier from the given index and arena
+    /// identifier.
+    ///
+    /// ## Panics
+    ///
+    /// Implementations are allowed to panic if the given index is larger than
+    /// the underlying storage (e.g. the implementation uses a `u8` for storing
+    /// indices and the given index value is larger than 255).
+    fn new_id(index: usize, arena_id: usize) -> Self::Id;
```
Comment (matklad) [inline src/lib.rs]: The order of arguments feels "swapped" here: arena is "big", I'd expect it to be first.

**Comment by matklad** [comment]
Heh, due to coherence, I've had to wrap both the index and the arena, which resulted in *more* lines of code then the basic implementation: https://github.com/rust-analyzer/rust-analyzer/commit/88cf81d653f615dd9d70d6b38c139c0455a7fdec

**Comment by fitzgen** [comment]
What upstream traits are you implementing for the arena?

**Comment by fitzgen** [comment]
Anything we can do to help cut down on boilerplate?

**Comment by matklad** [comment]
>What upstream traits are you implementing for the arena?

Excellent question! I was implementing `Index<MyId>`, but of course I don't need it: there's a blanket impl in id-arena. So now migration is a win: https://github.com/rust-analyzer/rust-analyzer/commit/0eb3b170dd96b107d7d89b9bfa252399e2bed6e9

**Comment by fitzgen** [comment]
Swapped the order of arena id and index for `new_id`. Once this is green, will merge it!

BTW, the `keys` and `values` methods would be good follow up PRs!

**Comment by matklad** [comment]
>BTW, the keys and values methods would be good follow up PRs!

Do we actually need them though? `.iter` is general enough, and the boilerplate to convert it to keys/values is pretty minimal. 

BTW, we don't have a random-access iterator in Rust, right? I suddenly realized that, because I allocate items in a particular order, I can binary search them if the underlying array impl were somehow exposed.

**Comment by fitzgen** [comment]
> Do we actually need them though? `.iter` is general enough, and the boilerplate to convert it to keys/values is pretty minimal.

Fair enough.

> BTW, we don't have a random-access iterator in Rust, right? I suddenly realized that, because I allocate items in a particular order, I can binary search them if the underlying array impl were somehow exposed.

Nope, just indexing.

---------------------

Any other changes you want to land before publishing a new breaking version?

**Comment by matklad** [comment]
* What do you think about making `arena_id` `u32`? This is not for perf, but to make writing `ArenaBehavior` easier: i've messed up `arena_id` and `index` twice :) If they were `u32` and `usize`, that would have helped. IIRC, we don't use `arena_id` as index, so, arguably, `usize` is a wrong choice anyway. 

* I haven't looked into adding `type Gen = ()` to `ArenaBehavior`, but, if we want to do that, we probably need to do it before breaking the API.

**Comment by fitzgen** [comment]
>     * I haven't looked into adding `type Gen = ()` to `ArenaBehavior`, but, if we want to do that, we probably need to do it before breaking the API.

I think it is safe to push this back to another release. If we *do* end up making `ArenaBehavior` generic enough to represent generational indices in a clean way, then it seems fine to me to do another breaking bump.

It isn't like there are that many users of this crate, after all ;) Mostly just some of my personal projects and your branch of rust-analyze, AFAIK.

**Comment by fitzgen** [comment]
>     * What do you think about making `arena_id` `u32`? This is not for perf, but to make writing `ArenaBehavior` easier: i've messed up `arena_id` and `index` twice :) If they were `u32` and `usize`, that would have helped. IIRC, we don't use `arena_id` as index, so, arguably, `usize` is a wrong choice anyway.

SGTM

**Pull request merged**